homeassistant/helpers/script.py

```diff
@@ -141,6 +141,11 @@ def __init__(
         self._hass = hass
         self._script = script
         self._variables = variables or {}
```

Comment:
If this isn't a dict, but a MappingProxyType, we need to adjust the typing of this parameter. When is it a `MappingProxyType`?

Reply:
This class can accept either a `dict` or a `MappingProxyType`. It doesn't really care. I didn't appreciate at first that it might get a `MappingProxyType` until I ran into this problem. So, yes, you're right, the typing info probably needs updating. (I'm actually not even sure what `TemplateVarsType` is. It was like that from before I started all this scripting work and never second guessed it.)

To answer your question, when a script is called via a service call, then the variables are passed in as a `MappingProxyType` (because they come from `ServiceCall.data`.) But when an automation's actions are run, then the variables are usually passed in as a `dict`.

But more to the point, I think I need to change two things: 1) make this change in `Script.async_run()` instead of here, and 2) always make a copy of the variables if `top_level` is true, regardless of type. These changes, now that I've thought about it some more, are not only more appropriate, they will also guarantee that there will be no "leakage" of variables created during a run getting accidentally returned to the caller. This is not as important now as for the `wait_for_trigger` action I'm working on next. Might as well do it now.

So I propose I commit another change per the above.

Reply:
Just looked up `TemplateVarsType `:
```python
TemplateVarsType = Optional[Dict[str, Any]]
```

Reply:
The suggestion sounds good, although I didn't follow the part about the leakage. I'm not very familiar with this code.

Reply:
Just as an FYI, during a script run, if that script contains a `repeat` action, then it creates a variable named `repeat`. When I add the `wait_for_trigger` action it will create a variable named `wait_trigger`. If the passed in variables container was a `dict`, and it was just used directly, then these variables, that are only supposed to exist within the script "run", could possibly get left in the `dict` and, hence, exist outside the script run context. So the thought is, not only does making a writable copy deal with the read-only nature of a `MappingProxyType`, it also guarantees any variables created during the run can't get back into the caller's context.

Reply:
Ok, thanks for clarifying that. :+1: